File: Android/app/src/main/java/com/diveapp/android/model/DiveLogModels.kt
package com.diveapp.android.model

import com.diveapp.android.core.network.LocalDateSerializer
import com.diveapp.android.core.network.OffsetDateTimeSerializer
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import java.time.LocalDate
import java.time.OffsetDateTime

@Serializable
data class DiveLocationInput(
    val name: String,
    val address: String? = null,
    val latitude: Double,
    val longitude: Double,
    @SerialName("naver_place_id") val naverPlaceId: String? = null,
    val country: String? = null,
    val city: String? = null,
)

@Serializable
data class DiveLocationResponse(
    val id: String,
    val name: String,
    val address: String?,
    val latitude: Double,
    val longitude: Double,
    val country: String?,
    val city: String?,
)

@Serializable
data class FreedivingDetailInput(
    @SerialName("max_depth") val maxDepth: Double,
    @SerialName("dive_time_seconds") val diveTimeSeconds: Int,
)

@Serializable
data class ScubaDetailInput(
    @SerialName("max_depth") val maxDepth: Double,
    @SerialName("dive_time_seconds") val diveTimeSeconds: Int,
    @SerialName("tank_pressure_start") val tankPressureStart: Int,
    @SerialName("tank_pressure_end") val tankPressureEnd: Int,
)

@Serializable
data class DivePhotoInput(
    @SerialName("image_url") val imageUrl: String,
    @SerialName("display_order") val displayOrder: Int = 1,
)

@Serializable
data class DiveLogCreateRequest(
    @SerialName("dive_type") val diveType: DiveType,
    @Serializable(with = LocalDateSerializer::class)
    @SerialName("dive_date") val diveDate: LocalDate,
    val location: DiveLocationInput,
    val latitude: Double? = null,
    val longitude: Double? = null,
    val memo: String? = null,
    val freediving: FreedivingDetailInput? = null,
    val scuba: ScubaDetailInput? = null,
    val photos: List<DivePhotoInput> = emptyList(),
)

@Serializable
data class DiveLogUpdateRequest(
    val memo: String? = null,
    val freediving: FreedivingDetailInput? = null,
    val scuba: ScubaDetailInput? = null,
)

@Serializable
data class FreedivingDetailResponse(
    @SerialName("max_depth") val maxDepth: Double,
    @SerialName("dive_time_seconds") val diveTimeSeconds: Int,
)

@Serializable
data class ScubaDetailResponse(
    @SerialName("max_depth") val maxDepth: Double,
    @SerialName("dive_time_seconds") val diveTimeSeconds: Int,
    @SerialName("tank_pressure_start") val tankPressureStart: Int,
    @SerialName("tank_pressure_end") val tankPressureEnd: Int,
)

@Serializable
data class DivePhotoResponse(
    val id: String,
    @SerialName("image_url") val imageUrl: String,
    @SerialName("display_order") val displayOrder: Int,
)

@Serializable
data class DiveLogResponse(
    val id: String,
    @SerialName("dive_type") val diveType: DiveType,
    @Serializable(with = LocalDateSerializer::class)
    @SerialName("dive_date") val diveDate: LocalDate,
    val location: DiveLocationResponse,
    val memo: String?,
    val freediving: FreedivingDetailResponse?,
    val scuba: ScubaDetailResponse?,
    val photos: List<DivePhotoResponse>,
    @Serializable(with = OffsetDateTimeSerializer::class)
    @SerialName("created_at") val createdAt: OffsetDateTime,
)

@Serializable
data class DiveLogListItem(
    val id: String,
    @SerialName("dive_type") val diveType: DiveType,
    @Serializable(with = LocalDateSerializer::class)
    @SerialName("dive_date") val diveDate: LocalDate,
    @SerialName("location_name") val locationName: String,
    @SerialName("cover_image_url") val coverImageUrl: String?,
    @SerialName("max_depth") val maxDepth: Double,
)

@Serializable
data class DiveLogStatisticsResponse(
    @SerialName("total_dive_count") val totalDiveCount: Int,
    @SerialName("max_depth_overall") val maxDepthOverall: Double,
    @SerialName("total_dive_time_seconds") val totalDiveTimeSeconds: Int,
    @SerialName("freediving_count") val freedivingCount: Int,
    @SerialName("scuba_count") val scubaCount: Int,
)

File: Web/src/models/diveLogModels.ts
import type { DiveType } from './enums';

export interface DiveLocationInput {
  name: string;
  address?: string | null;
  latitude: number;
  longitude: number;
  naver_place_id?: string | null;
  country?: string | null;
  city?: string | null;
}

export interface DiveLocationResponse {
  id: string;
  name: string;
  address: string | null;
  latitude: number;
  longitude: number;
  country: string | null;
  city: string | null;
}

export interface FreedivingDetailInput {
  max_depth: number;
  dive_time_seconds: number;
}

export interface ScubaDetailInput {
  max_depth: number;
  dive_time_seconds: number;
  tank_pressure_start: number;
  tank_pressure_end: number;
}

export interface DivePhotoResponse {
  id: string;
  image_url: string;
  display_order: number;
}

export interface DiveLogCreateRequest {
  dive_type: DiveType;
  dive_date: string;
  location: DiveLocationInput;
  // Exact coordinate selected for this individual log — kept separate from
  // `location`'s coordinate since the reusable DiveLocation row can be shared
  // (and later re-pointed) across dive logs with a similar name/city/country.
  latitude?: number | null;
  longitude?: number | null;
  memo?: string | null;
  freediving?: FreedivingDetailInput | null;
  scuba?: ScubaDetailInput | null;
}

export interface DiveLogUpdateRequest {
  memo?: string | null;
  freediving?: FreedivingDetailInput | null;
  scuba?: ScubaDetailInput | null;
}

export interface DiveLogResponse {
  id: string;
  dive_type: DiveType;
  dive_date: string;
  location: DiveLocationResponse;
  latitude: number | null;
  longitude: number | null;
  memo: string | null;
  freediving: FreedivingDetailInput | null;
  scuba: ScubaDetailInput | null;
  photos: DivePhotoResponse[];
  created_at: string;
}

/** A point chosen in the map picker (Docs/13), before it's attached to a
 * create/update request. UI-only — not sent to the API as-is. */
export interface PickedDiveLocation {
  siteName: string;
  latitude: number;
  longitude: number;
}

export interface DiveLogListItem {
  id: string;
  dive_type: DiveType;
  dive_date: string;
  location_name: string;
  cover_image_url: string | null;
  max_depth: number;
}

export interface DiveLogStatisticsResponse {
  total_dive_count: number;
  max_depth_overall: number;
  total_dive_time_seconds: number;
  freediving_count: number;
  scuba_count: number;
}
